src/app-layout/visual-refresh-toolbar/toolbar/trigger-button/index.tsx

```diff
@@ -44,11 +75,145 @@ function TriggerButton(
     badge,
     selected = false,
     highContrastHeader,
+    hasTooltip = false,
+    tooltipText,
+    hasOpenDrawer = false,
+    isMobile = false,
+    isForPreviousActiveDrawer = false,
+    isForSplitPanel = false,
   }: TriggerButtonProps,
   ref: React.Ref<ButtonProps.Ref>
 ) {
+  const containerRef = useRef<HTMLDivElement | null>(null);
+  const tooltipValue = tooltipText ? tooltipText : ariaLabel ? ariaLabel : '';
+  const [showTooltip, setShowTooltip] = useState<boolean>(false);
+  const [suppressTooltip, setSupressTooltip] = useState<boolean>(false);
+
+  const handleTriggerClick = (event: React.MouseEvent<HTMLButtonElement>) => {
+    event.stopPropagation(); // Stop the event from propagating to the badge icon
+    setShowTooltip(false);
+    setSupressTooltip(true);
+    onClick(event);
+  };
+
+  const handleBlur = (keepSupressed = false) => {
+    setSupressTooltip(keepSupressed);
+    setShowTooltip(false);
+  };
+
+  const handlePointerEnter = () => {
+    setSupressTooltip(false);
+    setShowTooltip(true);
+  };
+
+  /**
+   * Takes the drawer being closed and the data-shift-focus value from a close button on that drawer that persists
+   * on the event relatedTarget to determine not to show the tooltip
+   * @param event
+   */
+  const handleOnFocus = useCallback(
+    (event: FocusEvent) => {
+      const eventWithRelatedTarget = event as any;
+      const relatedTarget = eventWithRelatedTarget?.relatedTarget;
+      const isFromAnotherTrigger = relatedTarget?.dataset?.shiftFocus === 'awsui-layout-drawer-trigger';
+      let shouldShowTooltip = false;
+      if (isForSplitPanel) {
+        const breadcrumbsComponent = document.querySelector('nav[data-awsui-discovered-breadcrumbs="true"]');
+        if (breadcrumbsComponent && relatedTarget && breadcrumbsComponent.contains(relatedTarget)) {
+          //show tooltip when using keys to navigate to it from the breadcrumbs
+          shouldShowTooltip = true;
+        } else {
+          shouldShowTooltip = isFromAnotherTrigger;
+        }
+      } else if (!isForPreviousActiveDrawer) {
+        shouldShowTooltip = true; //for keyed navigation inside the toolbar
+      } else if (isFromAnotherTrigger) {
+        //needed for safari which doesn't read the relatedTarget when drawer closed via
+        //drawer close button
+        shouldShowTooltip = true;
+      }
+      setSupressTooltip(!shouldShowTooltip);
+      setShowTooltip(true);
+    },
+    [
+      // To assert reference equality check
+      isForPreviousActiveDrawer,
+      isForSplitPanel,
+    ]
+  );
+
+  const tooltipVisible = useMemo(() => {
```

Comment:
Minor: I don't think `useMemo` really has any value here. This is not an expensive calculation and this might be adding more computation (and complexity) than it saves.

Reply:
Oh I see, it's so that you can use it as a dependency in a hook further down (line 202)?